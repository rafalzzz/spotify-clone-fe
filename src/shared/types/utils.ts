export type TCustomRequest = {
  basicUrl?: string;
  endpoint?: string;
  method?: 'GET' | 'POST' | 'PUT' | 'DELETE';
  headers?: Record<string, string>;
  requestBody?: Record<string, unknown>;
  allowCookies?: boolean;
};

export type MockResizeObserver = {
  observe: jest.Mock;
  unobserve: jest.Mock;
  disconnect: jest.Mock;
};
