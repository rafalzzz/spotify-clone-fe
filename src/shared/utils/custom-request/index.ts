import https from 'https';

import fetch from 'node-fetch';

const agent = new https.Agent({
  rejectUnauthorized: false,
});

type CustomRequestProps = {
  basicUrl?: string;
  endpoint?: string;
  method?: 'GET' | 'POST' | 'PUT' | 'DELETE';
  headers?: Record<string, string>;
  requestBody?: Record<string, unknown>;
};

export const customRequest = ({
  basicUrl = process.env.NEXT_PUBLIC_API_URL,
  endpoint = '/',
  method,
  headers = {
    Accept: 'application/json',
    'Content-Type': 'application/json',
  },
  requestBody,
}: CustomRequestProps) =>
  fetch(basicUrl + endpoint, {
    method,
    headers,
    agent,
    body: JSON.stringify(requestBody),
  });
