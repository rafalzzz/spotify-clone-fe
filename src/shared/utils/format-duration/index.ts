const pad = (number: number): string => {
  return ('0' + number).slice(-2);
};

export const formatDuration = (seconds: number, isReversedTime: boolean = false): string => {
  if (seconds < 0) return '0:00';

  const date = new Date(seconds * 1000);
  const hh = date.getUTCHours();
  const mm = date.getUTCMinutes();
  const ss = pad(date.getUTCSeconds());

  let value = `${mm}:${ss}`;

  if (hh) {
    value = `${hh}:${pad(mm)}:${ss}`;
  }

  if (isReversedTime) {
    return `-${value}`;
  }

  return value;
};
