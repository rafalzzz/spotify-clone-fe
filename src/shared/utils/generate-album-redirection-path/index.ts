export const generateAlbumRedirectionPath = (artistName: string): string =>
  `/album/${artistName.toLocaleLowerCase()}`;
