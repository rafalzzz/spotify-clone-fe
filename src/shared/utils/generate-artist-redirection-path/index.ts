export const generateArtistRedirectionPath = (artistName: string): string =>
  `/artist/${artistName.toLocaleLowerCase()}`;
