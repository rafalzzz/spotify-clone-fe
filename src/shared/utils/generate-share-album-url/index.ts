export const generateShareAlbumUrl = (albumId: number): string =>
  window.location.origin + `/album/${albumId}`;
