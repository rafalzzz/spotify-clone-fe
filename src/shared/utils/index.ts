export { convertMusicTrackToSongItem } from './convert-music-track-to-song-item';
export { getItemsAmount } from './get-items-amount';
export { getMainArtist } from './get-main-artist';
export { generateAlbumRedirectionPath } from './generate-album-redirection-path';
export { generateArtistRedirectionPath } from './generate-artist-redirection-path';
export { generateShareTrackUrl } from './generate-share-track-url';
export { generateTrackRedirectionPath } from './generate-track-redirection-path';
export { getShuffledIndexes } from './get-shuffled-indexes';
