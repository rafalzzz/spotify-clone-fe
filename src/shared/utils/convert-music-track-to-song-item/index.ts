import { TSongItem } from '@/types/components';
import { EMusicTrackKeys, TMusicTrack } from '@/types/music-track';

export const convertMusicTrackToSongItem = (musicTrack: TMusicTrack): TSongItem => ({
  [EMusicTrackKeys.ARTIST_NAME]: musicTrack[EMusicTrackKeys.ARTIST_NAME],
  [EMusicTrackKeys.COLLECTION_NAME]: musicTrack[EMusicTrackKeys.COLLECTION_NAME],
  [EMusicTrackKeys.TRACK_NAME]: musicTrack[EMusicTrackKeys.TRACK_NAME],
  [EMusicTrackKeys.PREVIEW_URL]: musicTrack[EMusicTrackKeys.PREVIEW_URL],
  [EMusicTrackKeys.ARTWORK_URL_60]: musicTrack[EMusicTrackKeys.ARTWORK_URL_60],
  [EMusicTrackKeys.ARTIST_ID]: musicTrack[EMusicTrackKeys.ARTIST_ID],
  [EMusicTrackKeys.COLLECTION_ID]: musicTrack[EMusicTrackKeys.COLLECTION_ID],
  [EMusicTrackKeys.TRACK_ID]: musicTrack[EMusicTrackKeys.TRACK_ID],
});
