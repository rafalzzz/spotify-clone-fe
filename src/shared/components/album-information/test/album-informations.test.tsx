import { render, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import { useRouter } from 'next/navigation';

import { generateArtistRedirectionPath } from '@/utils/generate-artist-redirection-path';

import { AlbumInformation } from '..';

const MOCKED_COLLECTION_NAME = 'Mocked album name';
const MOCKED_RELEASE_DATE = '2020-01-01';
const MOCKED_ARTIST_NAME = `Mocked artist name`;

const COLLECTION_NAME_TEST_ID = 'album-information-collection-name';
const RELEASE_DATE_TEST_ID = 'album-information-release-date';
const ARTIST_NAME_TEST_ID = 'album-information-artist-name';

const TRACK_NAME_TOOLTIP_TEST_ID = 'album-information-collection-name-tooltip';
const ARTIST_NAME_TOOLTIP_TEST_ID = 'album-information-artist-name-tooltip';

const mockPush = jest.fn();

beforeEach(() => {
  (useRouter as jest.Mock).mockImplementation(() => ({
    push: mockPush,
  }));
});

const renderAlbumInformation = () =>
  render(
    <AlbumInformation
      collectionName={MOCKED_COLLECTION_NAME}
      releaseDate={MOCKED_RELEASE_DATE}
      artistName={MOCKED_ARTIST_NAME}
    />,
  );

describe('AlbumInformation', () => {
  it('renders component without error', () => {
    const screen = renderAlbumInformation();
    expect(screen).toMatchSnapshot();
  });

  it('renders the collection name, release year and artist name', () => {
    const { queryByTestId } = renderAlbumInformation();

    const collectionNameElement = queryByTestId(COLLECTION_NAME_TEST_ID);
    const releaseDateElement = queryByTestId(RELEASE_DATE_TEST_ID);
    const artistNameElement = queryByTestId(ARTIST_NAME_TEST_ID);

    expect(collectionNameElement).toBeInTheDocument();
    expect(releaseDateElement).toBeInTheDocument();
    expect(artistNameElement).toBeInTheDocument();
  });

  it('passes collection name to the collection tooltip', () => {
    const { queryByTestId, queryByText } = renderAlbumInformation();

    const trackNameTooltip = queryByTestId(TRACK_NAME_TOOLTIP_TEST_ID);
    expect(trackNameTooltip).toBeInTheDocument();

    const tooltipTitle = queryByText(MOCKED_COLLECTION_NAME, { selector: 'span' });
    expect(tooltipTitle).toBeInTheDocument();
  });

  it('renders correct release year', () => {
    const { queryByTestId } = renderAlbumInformation();

    const releaseDateElement = queryByTestId(RELEASE_DATE_TEST_ID);
    expect(releaseDateElement).toHaveTextContent('2020');
  });

  it('passes artist name to the artist tooltip', () => {
    const { queryByTestId } = renderAlbumInformation();

    const artistNameTooltip = queryByTestId(ARTIST_NAME_TOOLTIP_TEST_ID);
    expect(artistNameTooltip).toBeInTheDocument();

    const tooltipTitle = (artistNameTooltip as HTMLElement).querySelector('span');
    expect(tooltipTitle).toBeInTheDocument();
  });

  it('navigates to the correct path when artist name is clicked', () => {
    const { queryByTestId } = renderAlbumInformation();
    const artistNameElement = queryByTestId(ARTIST_NAME_TEST_ID);

    fireEvent.click(artistNameElement as Element);

    const expectedPath = generateArtistRedirectionPath(MOCKED_ARTIST_NAME);
    expect(mockPush).toHaveBeenCalledWith(expectedPath);
  });
});
