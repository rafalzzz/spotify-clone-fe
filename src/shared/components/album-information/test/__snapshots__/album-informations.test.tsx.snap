// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AlbumInformation renders component without error 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="album-information"
      >
        <div
          data-mouseenterdelay="0.5"
          data-placement="right"
          data-testid="album-information-collection-name-tooltip"
        >
          <h3
            class="album-information__text"
            data-testid="album-information-collection-name"
          >
            Mocked album name
          </h3>
          <span>
            Mocked album name
          </span>
        </div>
        <div
          class="album-information__wrapper"
        >
          <span
            class="album-information__release-date"
            data-testid="album-information-release-date"
          >
            2020
          </span>
          <span
            class="album-information__separator"
          >
            •
          </span>
          <button
            class="album-information__text album-information__artist"
            data-testid="album-information-artist-name"
          >
            <div
              data-mouseenterdelay="0.5"
              data-placement="right"
              data-testid="album-information-artist-name-tooltip"
            >
              <span>
                Mocked artist name
              </span>
              <span>
                Mocked artist name
              </span>
            </div>
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="album-information"
    >
      <div
        data-mouseenterdelay="0.5"
        data-placement="right"
        data-testid="album-information-collection-name-tooltip"
      >
        <h3
          class="album-information__text"
          data-testid="album-information-collection-name"
        >
          Mocked album name
        </h3>
        <span>
          Mocked album name
        </span>
      </div>
      <div
        class="album-information__wrapper"
      >
        <span
          class="album-information__release-date"
          data-testid="album-information-release-date"
        >
          2020
        </span>
        <span
          class="album-information__separator"
        >
          •
        </span>
        <button
          class="album-information__text album-information__artist"
          data-testid="album-information-artist-name"
        >
          <div
            data-mouseenterdelay="0.5"
            data-placement="right"
            data-testid="album-information-artist-name-tooltip"
          >
            <span>
              Mocked artist name
            </span>
            <span>
              Mocked artist name
            </span>
          </div>
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
